src/dune/file_tree.ml

```diff
@@ -306,76 +370,123 @@ module rec Memoized : sig
+       mentioned in the dune file via [(chdir ..)]. We must handle their status
+       explicitly, but they do contribute to the visited directories map as they
+       can't cause cycles *)
+    let virtual_ ~sub_dirs ~parent_status ~dune_file ~init ~path =
+      match dune_file with
+      | None
+      | Some (Dune_file.Ocaml_script _) ->
+        init
+      | Some (Plain dune_file) ->
+        let dirs = Sub_dirs.Dir_map.sub_dirs dune_file.for_subdirs in
+        let status_map = Sub_dirs.eval sub_dirs ~dirs in
+        List.fold_left dirs ~init ~f:(fun acc fn ->
+            let path = Path.Source.relative path fn in
+            match status ~status_map ~parent_status fn with
+            | None -> acc
+            | Some dir_status ->
+              String.Map.update acc fn ~f:(function
+                (* Directories which are physical are skipped *)
+                | Some _ as r -> r
+                | None -> Some (make_subdir ~dir_status ~virtual_:true path)))
+
+    let all ~dirs_visited ~dirs ~sub_dirs ~parent_status ~dune_file ~path =
+      let visited, init =
+        physical ~dirs_visited ~dirs ~sub_dirs ~parent_status
+      in
+      let init = virtual_ ~sub_dirs ~parent_status ~dune_file ~init ~path in
+      (visited, init)
+  end
 
   let dune_file ~(dir_status : Sub_dirs.Status.t) ~recognize_jbuilder_projects
       ~path ~files ~project =
-    if dir_status = Data_only then
-      None
-    else if
-      (not recognize_jbuilder_projects)
-      && String.Set.mem files Dune_file.jbuild_fname
-    then
-      User_error.raise
-        ~loc:
-          (Loc.in_file
-             (Path.source (Path.Source.relative path Dune_file.jbuild_fname)))
-        [ Pp.text
-            "jbuild files are no longer supported, please convert this file to \
-             a dune file instead."
-        ; Pp.text
-            "Note: You can use \"dune upgrade\" to convert your project to \
-             dune."
-        ]
-    else if not (String.Set.mem files Dune_file.fname) then
+    let file_exists =
+      if dir_status = Data_only then
+        false
+      else if
+        (not recognize_jbuilder_projects)
+        && String.Set.mem files Dune_file.jbuild_fname
+      then
+        User_error.raise
+          ~loc:
+            (Loc.in_file
+               (Path.source (Path.Source.relative path Dune_file.jbuild_fname)))
+          [ Pp.text
+              "jbuild files are no longer supported, please convert this file \
+               to a dune file instead."
+          ; Pp.text
+              "Note: You can use \"dune upgrade\" to convert your project to \
+               dune."
+          ]
+      else
+        String.Set.mem files Dune_file.fname
+    in
+    let parent =
+      let open Option.O in
+      let* parent = Path.Source.parent path in
+      let* parent = find_dir parent in
+      match parent.contents.dune_file with
+      | None
+      | Some (Ocaml_script _) ->
+        None (* wrong but easy *)
```

Comment:
Can we forbid chdir in OCaml scripts so this is no longer wrong?

Reply:
I think it's already forbidden. The error should be the same as trying to output a subdir related stanza like `data_only_dirs`.